AGTRAP.py

```diff
@@ -0,0 +1,244 @@
+           default = (0, 2 ** 32)
+    :param quality_threshold:average sequence quality threshold for filtering (Phred33 scale);
+           sequences with quality below the threshold are discarded
+    :param report:report = True (show filtering report), report = True (default)
+    :return:A filtered fastq file containing sequences
+             that have passed all tests (according to specified parameters)
+    """
+
+    if not isinstance(gc_bounds, tuple):
+        gc_bounds = (0, gc_bounds)
+    if not isinstance(length_bounds, tuple):
+        length_bounds = (0, length_bounds)
+    if output_filename is None:
+        output_filename = os.path.splitext(os.path.basename(input_path))[0]
+    output_filename = f'{output_filename}.fastq'
+
+    os.makedirs('fastq_filtrator_resuls', exist_ok=True)
+    output_dir = 'fastq_filtrator_resuls'
+
+    count_filtered_record = 0
+    count_record = 0
+
+    with open(os.path.join(output_dir, output_filename), mode='w') as out_file:
+        for record in SeqIO.parse("reads.3.fastq", "fastq"):
+            count_record += 1
+            if gc_bounds[0] <= gc_fraction(record.seq)*100 <= gc_bounds[1]:
+                if length_bounds[0] <= len(record.seq) <= length_bounds[1]:
+                    if np.mean(record.letter_annotations['phred_quality']) > quality_threshold:
+                        count_filtered_record += 1
+                        SeqIO.write(record, out_file, "fastq")
+
+    if report:
+        print(f' Number of source sequences = {count_record},'
+              f'\n Number of sequences after filtering = {count_filtered_record},')
+
+    return f'{output_filename} file created!'
+
+
+class BiologicalSequence(ABC, str):
+
+    def __init__(self, seq: str):
+        super().__init__()
+        self.seq = seq
+
+    @abstractmethod
+    def __len__(self):
+        """Return the length of the sequence"""
+        return super().__len__()
+
+    @abstractmethod
+    def __getitem__(self, key: Union[int, slice]) -> object:
+        """Returnы a subsequence in the form of a single letter or a string of several letters"""
+        if isinstance(key, int):
+            if 0 >= key or key > len(self.seq):
+                raise IndexError(f"Index {key} is out of range of the sequence")
+            return super().__getitem__(key - 1)
+        elif isinstance(key, slice):
+            if key.start <= 0 or key.stop > len(self.seq):
+                raise IndexError(f"Index {key} is out of range of the sequence")
+            new_slice = slice(key.start - 1, key.stop, key.step)
+            return super().__getitem__(new_slice)
+
+    @abstractmethod
+    def _check_alphabet(self, seq: str):
+        """Checks the sequence to match the alphabet"""
+        pass
+
+    def __str__(self):
+        return f'{self.__class__.__name__}("{self.seq}")'
+
+
+class NucleicAcidSequence(BiologicalSequence):
+    alphabet = set('ATCGUatcgu')
+    compliment_dict = None

```

Comment:
Интересно, зачем тут алфавит, если он переопределяется дочерними классами? Хотя алфавит, конечно, логичный)
```suggestion
    alphabet = None
    compliment_dict = None
```